### PR TITLE
[SDF-65] Basic window and application functionality

### PR DESCRIPTION
**Pull request by kubazuch**
1. Even framework (to be enhanced in SDF-73)
2. Basic window and app functionality
3. boilerplate galore

### CODE REVIEW

File: resin/resin/event/window_events.hpp [hunk 1]
```diff
@@ -0,0 +1,39 @@
+#ifndef WIDNOW_EVENTS_HPP
```
Comment (adamieck) [inline resin/resin/event/window_events.hpp:1]: WIDNOW_EVENTS 😭
image

Comment (kubazuch) [reply resin/resin/event/window_events.hpp:1]: crying shaking throwing up

**Review by migoox — COMMENTED** [review]
Fantastic job, nitpicks only
image

File: resin/resin/core/graphics_context.cpp [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+void GraphicsContext::init() {
+  glfwMakeContextCurrent(window_ptr_);
+  const int glad_version = gladLoadGL(glfwGetProcAddress);
+  if (glad_version == 0) {
+    throw std::runtime_error("Failed to initialize OpenGL context");
+  }
+
+  resin::Logger::debug("GLAD version: {0}.{1}", GLAD_VERSION_MAJOR(glad_version), GLAD_VERSION_MINOR(glad_version));
+  resin::Logger::debug("OpenGL info:");
```
[the remark this answers is not in the document]
Comment (kubazuch) [reply resin/resin/core/graphics_context.cpp:25]: Done

File: resin/resin/core/window.hpp [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  Window& operator=(const Window&) = delete;
+  Window& operator=(Window&&)      = delete;
+
+ private:
+  static void api_init();
+  static void api_terminate();
+
+  void set_glfw_callbacks() const;
+
+ private:  // NOLINT
```
[the remark this answers is not in the document]
Comment (kubazuch) [reply resin/resin/core/window.hpp:62]: Done

File: resin/resin/resin.hpp [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+  void run();
+  void update(duration_t delta);
+  void render();
+
+  bool on_window_close(WindowCloseEvent& e);
+  bool on_window_resize(WindowResizeEvent& e);
+
+ public:
+  static constexpr duration_t kTickTime = 16666us;  // 60 TPS = 16.6(6) ms/t
```
[the remark this answers is not in the document]
Comment (migoox) [reply resin/resin/resin.hpp:43]: I misunderstood it then😅

File: resin/resin/core/window.hpp [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+  std::string title;
+  int x, y;
+  unsigned int width, height;
+  bool vsync, fullscreen;  // TODO(SDF-72): proper fullscreen handling
+
+  std::optional<std::reference_wrapper<EventDispatcher>> eventDispatcher;
+
+  explicit WindowProperties(std::string_view title = "Resin", unsigned int width = 1280U,             // NOLINT
+                            unsigned int height = 720U, bool vsync = false, bool fullscreen = false)  // NOLINT
+      : title(std::move(title)), width(width), height(height), vsync(vsync), fullscreen(fullscreen) {}
```
[the remark this answers is not in the document]
Comment (migoox) [reply resin/resin/core/window.hpp:27]: I think that simple `x=0`, `y=0` in the initializer list would be enough as if someone forgets to set the values after struct creation the `(0,0)` position is still a reasonable value. The point is that currently the values have no explicit defaults which produces undefined behavior when someone will construct the struct and refer to the `x` or `y`.

I've read that fortunately the `wuninitialized` flag should notice that a programmer has tried to read `x` or `y` when uninitialized, so you can ignore this suggestion.

**Review by migoox — APPROVED** [review]

**Pull request merged**